### PR TITLE
feat(attention): add deterministic CP reference

### PR DESCRIPTION
**Pull request by inaniloquentee**
## Scope
PR3 provides the independent deterministic CP Attention reference for prefill and chunked-prefill.

## Contract
| Area | Enforced behavior |
|---|---|
| Partial state | Materializes FP32 Out/LSE states |
| Merge order | Sorts logical global_block_index before every merge, independent of arrival order |
| Split-KV | Disabled in strict mode; CP/chunk layout cannot change arithmetic order |
| Inputs | Uses attention-ready Q/K/V with global RoPE positions, causal masks, varlen, padding, empty rows, and GQA |
| TE relation | TE may be used as an oracle; the reference remains the numerical contract |

## Validation
| Check | Result |
|---|---|
| Focused deterministic-reference suite | 46 passed |
| Integrated 8xH100 forward gate | Passed: `Out`/`LSE` max abs `0.0` across 2/4/8 ranks |

## Boundary
This PR validates reference arithmetic and fixed logical merge semantics. It does not claim real NCCL/AG/RS execution, fused CUDA arithmetic, decode replay, or performance.

Commit: 988d954

### CODE REVIEW

File: rl_engine/kernels/ops/pytorch/attention/cp_attention.py [hunk 1]
```diff
@@ -0,0 +1,548 @@
...
+
+
+CPAttentionReferenceOp = DeterministicCPAttentionReferenceOp
+
+__all__ = [
+    "AttentionPartialState",
+    "CPAttentionReferenceOp",
+    "DeterministicCPAttentionReferenceOp",
+    "merge_attention_partial_states",
+]
```
Comment (zhangj1an) [inline rl_engine/kernels/ops/pytorch/attention/cp_attention.py]: delete the alias `DeterministicCPAttentionReferenceOp` please

**Review by zhangj1an — APPROVED** [review]
LGTM. Thanks for your work!

This FP32 Ring CP golden reference is numerically equivalent with the PyTorch implementation of `flash_attn_fwd_softmax_lse_correction / flash_attn_fwd_out_correction`  in Transformer Engine, https://github.com/NVIDIA/TransformerEngine/blob/main/transformer_engine/pytorch/attention/dot_product_attention/context_parallel.py. The softmax merging order is by KV block order. 

Also the unit tests covered both forward and backward test cases (gradients via autograd).

**Comment by Flink-ddd** [comment]
please resolve the code conflicts first, Thank you.

**Comment by inaniloquentee** [comment]
> LGTM. Thanks for your work!
> 
> This FP32 Ring CP golden reference is numerically equivalent with the PyTorch implementation of `flash_attn_fwd_softmax_lse_correction / flash_attn_fwd_out_correction` in Transformer Engine, https://github.com/NVIDIA/TransformerEngine/blob/main/transformer_engine/pytorch/attention/dot_product_attention/context_parallel.py. The softmax merging order is by KV block order.
> 
> Also the unit tests covered both forward and backward test cases (gradients via autograd).

Thanks for the review! Removed the alias and kept only DeterministicCPAttentionReferenceOp.

Also agreed on the TE point: this PR keeps the PyTorch golden reference self-contained, while matching the same FP32 (out, lse) correction semantics as Transformer Engine.

**Comment by inaniloquentee** [comment]
> code conflicts

Fixed code conflicts.

**Review by Flink-ddd — APPROVED** [review]
LGTM
